File: app/assets/javascripts/game/PortfolioController.js
(function () {
    var app = angular.module('shocktrade');

    /**
     * Portfolio Controller
     */
    app.controller('PortfolioController', ['$scope', '$log', '$timeout', 'toaster', 'MySession', 'ContestService', 'NewOrderDialog', 'QuoteService',
        function ($scope, $log, $timeout, toaster, MySession, ContestService, NewOrderDialog, QuoteService) {

            $scope.selectedClosedOrder = null;
            $scope.selectedOrder = null;
            $scope.selectedPosition = null;
            $scope.selection = {
                exposure: "sector",
                performance: "gains"
            };

            $scope.portfolioTabs = [{
                "name": "Positions",
                "imageURL": "/assets/images/objects/position.png",
                "path": "/assets/views/play/portfolio/positions.htm",
                "active": false,
                "init": function (contest) {
                    // TODO enrich positions here
                },
                "isLocked": function (contest) {
                    return contest.status !== 'ACTIVE';
                }
            }, {
                "name": "Open Orders",
                "imageURL": "/assets/images/objects/portfolio_query.png",
                "path": "/assets/views/play/portfolio/orders_active.htm",
                "active": false,
                "init": function (contest) {
                    // TODO enrich orders here
                },
                "isLocked": function (contest) {
                    return contest.status !== 'ACTIVE';
                }
            }, {
                "name": "Closed Orders",
                "imageURL": "/assets/images/objects/portfolio_header.png",
                "path": "/assets/views/play/portfolio/orders_closed.htm",
                "active": false,
                "init": function (contest) {
                    // TODO enrich orders here
                },
                "isLocked": function (contest) {
                    return contest.status !== 'ACTIVE';
                }
            }, {
                "name": "Messages",
                "imageURL": "/assets/images/objects/chat.png",
                "path": "/assets/views/play/portfolio/chat.htm",
                "active": false,
                "init": function (contest) {
                    // TODO enrich orders here
                },
                "isLocked": function (contest) {
                    return contest.status !== 'ACTIVE';
                }
            }];

            /////////////////////////////////////////////////////////////////////
            //          Participant Functions
            /////////////////////////////////////////////////////////////////////

            $scope.getParticipant = function (contest) {
                if ($scope.participant == null) {
                    $scope.participant = $scope.findPlayerByID(contest, MySession.getUserID());
                    enrichParticipant(contest, $scope.participant);
                }
                return $scope.participant;
            };

            function enrichParticipant(contest, participant) {
                // enrich the orders
                ContestService.getEnrichedOrders(contest.OID(), participant.OID())
                    .success(function (response) {
                        var enrichedOrders = response;
                        $log.info("Loaded enriched orders " + angular.toJson(enrichedOrders));
                        participant.orders = enrichedOrders;
                    })
                    .error(function (err) {
                        toaster.pop('error', 'Error!', "Error loading enriched orders");
                    });

                // enrich the positions
                ContestService.getEnrichedPositions(contest.OID(), participant.OID())
                    .success(function (response) {
                        var enrichedPositions = response;
                        $log.info("Loaded enriched positions " + angular.toJson(enrichedPositions));
                        participant.positions = enrichedPositions;
                    })
                    .error(function (err) {
                        toaster.pop('error', 'Error!', "Error loading enriched positions");
                    });
            }

            /////////////////////////////////////////////////////////////////////
            //          Order Functions
            /////////////////////////////////////////////////////////////////////

            $scope.getActiveOrders = function (contest) {
                var participant = $scope.getParticipant(contest);
                return participant ? (participant.orders || []) : [];
            };

            $scope.getClosedOrders = function (contest) {
                var participant = $scope.getParticipant(contest);
                return participant ? (participant.orderHistory || []) : [];
            };

            $scope.popupNewOrderDialog = function () {
                NewOrderDialog.popup({symbol: QuoteService.lastSymbol});
            };

            $scope.isOrderSelected = function () {
                return $scope.selectedOrder != null;
            };

            $scope.selectOrder = function (order) {
                $scope.selectedOrder = order;
            };

            $scope.toggleSelectedOrder = function () {
                $scope.selectedOrder = null;
            };

            $scope.isClosedOrderSelected = function () {
                return $scope.selectedClosedOrder != null;
            };

            $scope.selectClosedOrder = function (order) {
                $scope.selectedClosedOrder = order;
            };

            $scope.toggleSelectedClosedOrder = function () {
                $scope.selectedClosedOrder = null;
            };

            /////////////////////////////////////////////////////////////////////
            //          Position Functions
            /////////////////////////////////////////////////////////////////////

            $scope.getPositions = function (contest) {
                var participant = $scope.getParticipant(contest);
                return participant ? (participant.positions || []) : [];
            };


            $scope.isPositionSelected = function () {
                return $scope.selectedPosition != null;
            };

            $scope.selectPosition = function (position) {
                $scope.selectedPosition = position;
            };

            $scope.sellPosition = function (symbol, quantity) {
                NewOrderDialog.popup({symbol: symbol, quantity: quantity});
            };

            $scope.toggleSelectedPosition = function () {
                $scope.selectedPosition = null;
            };

            /////////////////////////////////////////////////////////////////////
            //          Summary Functions
            /////////////////////////////////////////////////////////////////////

            $scope.getCashAvailable = function (contest) {
                var participant = $scope.getParticipant(contest);
                return participant ? (participant.fundsAvailable || 0) : 0;
            };

            $scope.asOfDate = function (contest) {
                var participant = $scope.getParticipant(contest);
                return participant && participant.lastTradeTime ? participant.lastTradeTime : new Date();
            };

            $scope.getTotalOrders = function (contest) {
                return $scope.getTotalBuyOrders(contest) + $scope.getTotalSellOrders(contest);
            };

            $scope.getTotalEquity = function (contest) {
                return $scope.getTotalInvestment(contest) + $scope.getCashAvailable(contest);
            };

            $scope.getTotalInvestment = function (contest) {
                var participant = $scope.getParticipant(contest);
                var total = 0;
                if (participant != null) {
                    angular.forEach(participant.positions, function (p) {
                        total += p.netValue;
                    });
                }
                return total;
            };

            $scope.getTotalBuyOrders = function (contest) {
                var participant = $scope.getParticipant(contest);
                var total = 0;
                if (participant != null) {
                    angular.forEach(participant.orders, function (o) {
                        if (o.orderType == 'BUY') {
                            total += o.price * o.quantity + o.commission;
                        }
                    });
                }
                return total;
            };

            $scope.getTotalSellOrders = function (contest) {
                var participant = $scope.getParticipant(contest);
                var total = 0;
                if (participant != null) {
                    angular.forEach(participant.orders, function (o) {
                        if (o.orderType == 'SELL') {
                            total += o.price * o.quantity + o.commission;
                        }
                    });
                }
                return total;
            };

            $scope.isMarketOrder = function (order) {
                return order.priceType == 'MARKET' || order.priceType == 'MARKET_ON_CLOSE';
            };

            $scope.tradingStart = function () {
                return (new Date()).getTime(); // TODO replace with service call
            };

            /////////////////////////////////////////////////////////////////////
            //          Chart Functions
            /////////////////////////////////////////////////////////////////////

            $scope.getExposureChart = function (target) {
                $scope.getChart(MySession.getContestID(), MySession.getUserName(), $scope.selection.exposure, target);
            };

            $scope.getPerformanceChart = function (target) {
                $scope.getChart(MySession.getContestID(), MySession.getUserName(), $scope.selection.performance, target);
            };

            $scope.getChart = function (contestId, participantName, chartName, target) {
                if ((contestId && contestId != "") && (participantName && participantName != "")) {
                    // load the chart representing the securities
                    ContestService.getChart(contestId, participantName, chartName).then(
                        function (exposureData) {
                            // create the chart title & sub-title
                            var subTitle = chartName.capitalize();
                            var title = (chartName == "gains" || chartName == "losses") ? "Performance " + subTitle : subTitle + " Exposure";

                            // construct the chart
                            var exposure = new AmCharts.AmPieChart();
                            exposure.colors = ["#00dd00", "#ff00ff", "#00ffff", "#885500", "#0044ff", "#888800"];
                            exposure.addTitle(title, 12);
                            exposure.dataProvider = exposureData;
                            exposure.titleField = "title";
                            exposure.valueField = "value";
                            exposure.sequencedAnimation = true;
                            exposure.startEffect = "elastic";
                            exposure.startDuration = 2;
                            exposure.innerRadius = "30%";
                            exposure.labelRadius = 15;

                            // 3D chart
                            exposure.depth3D = 10;
                            exposure.angle = 15;
                            exposure.write(target);
                        },
                        function (err) {
                            $log.info("Error: " + angular.toJson(err));
                        });
                }
            };

            //////////////////////////////////////////////////////////////////////
            //              Watch Event Listeners
            //////////////////////////////////////////////////////////////////////

            function reset() {
                $scope.participant = null;
                $scope.selectedClosedOrder = null;
                $scope.selectedOrder = null;
                $scope.selectedPosition = null;
            }

            /**
             * Listen for contest update events
             */
            $scope.$on("contest_updated", function (event, contest) {
                $log.info("[Portfolio] Contest '" + contest.name + "' updated");
                reset();
            });

            /**
             * Watch for contest change events
             */
            $scope.$watch(MySession.contest, function () {
                reset();
            }, true);

        }]);

})();

File: app-web-angularjs/src/main/scala/com/shocktrade/javascript/dashboard/PortfolioController.scala
package com.shocktrade.javascript.dashboard

import com.shocktrade.javascript.AppEvents._
import com.shocktrade.javascript.dialogs.NewOrderDialogController.NewOrderDialogResult
import com.shocktrade.javascript.dialogs.{NewOrderDialog, NewOrderParams}
import com.shocktrade.javascript.discover.DiscoverController
import com.shocktrade.javascript.models.contest._
import com.shocktrade.javascript.{GlobalLoading, MySessionService}
import org.scalajs.angularjs.cookies.Cookies
import org.scalajs.angularjs.toaster.Toaster
import org.scalajs.angularjs.{Controller, Scope, Timeout, injected}
import org.scalajs.dom.console
import org.scalajs.nodejs.util.ScalaJsHelper._

import scala.scalajs.concurrent.JSExecutionContext.Implicits.queue
import scala.scalajs.js
import scala.scalajs.js.annotation.ScalaJSDefined
import scala.util.{Failure, Success}

/**
  * Portfolio Controller
  * @author lawrence.daniels@gmail.com
  */
class PortfolioController($scope: PortfolioScope, $cookies: Cookies, $timeout: Timeout, toaster: Toaster,
                          @injected("MySessionService") mySession: MySessionService,
                          @injected("ContestService") contestService: ContestService,
                          @injected("NewOrderDialog") newOrderDialog: NewOrderDialog)
  extends Controller with GlobalLoading {

  private val marketOrderTypes = js.Array("MARKET", "MARKET_ON_CLOSE")

  $scope.selectedClosedOrder = js.undefined
  $scope.selectedOrder = js.undefined
  $scope.selectedPosition = js.undefined

  $scope.portfolioTabs = js.Array(
    new PortfolioTab(name = "Chat", icon = "fa-comment-o", path = "/assets/views/dashboard/chat.htm", active = false),
    new PortfolioTab(name = "Positions", icon = "fa-list-alt", path = "/assets/views/dashboard/positions.htm", active = false),
    new PortfolioTab(name = "Open Orders", icon = "fa-folder-open-o", path = "/assets/views/dashboard/orders_active.htm", active = false),
    new PortfolioTab(name = "Closed Orders", icon = "fa-folder-o", path = "/assets/views/dashboard/orders_closed.htm", active = false),
    new PortfolioTab(name = "Performance", icon = "fa-bar-chart-o", path = "/assets/views/dashboard/performance.htm", active = false),
    new PortfolioTab(name = "Exposure", icon = "fa-pie-chart", path = "/assets/views/dashboard/exposure.htm", active = false))

  /////////////////////////////////////////////////////////////////////
  //          Closed Order Functions
  /////////////////////////////////////////////////////////////////////

  $scope.getClosedOrders = () => {
    mySession.getClosedOrders.filter(_.accountType.contains($scope.getAccountType()))
  }

  $scope.isClosedOrderSelected = () => {
    $scope.getClosedOrders().nonEmpty && $scope.selectedClosedOrder.nonEmpty
  }

  $scope.selectClosedOrder = (closeOrder: js.UndefOr[ClosedOrder]) => {
    $scope.selectedClosedOrder = closeOrder
  }

  $scope.toggleSelectedClosedOrder = () => {
    $scope.selectedClosedOrder = js.undefined
  }

  /////////////////////////////////////////////////////////////////////
  //          Active Order Functions
  /////////////////////////////////////////////////////////////////////

  $scope.computeOrderCost = (anOrder: js.UndefOr[Order]) => anOrder.flatMap(_.totalCost)

  $scope.cancelOrder = (aContestId: js.UndefOr[String], aPlayerId: js.UndefOr[String], anOrderId: js.UndefOr[String]) => {
    for {
      contestId <- aContestId
      playerId <- aPlayerId
      orderId <- anOrderId
    } {
      asyncLoading($scope)(contestService.deleteOrder(contestId, playerId, orderId)) onComplete {
        case Success(contest) => mySession.setContest(contest)
        case Failure(err) =>
          toaster.error("Failed to cancel order")
      }
    }
  }

  $scope.getActiveOrders = () => {
    mySession.getOrders filter (_.accountType == $scope.getAccountType())
  }

  $scope.isMarketOrder = (anOrder: js.UndefOr[Order]) => {
    anOrder.exists(order => order.priceType.exists(marketOrderTypes.contains))
  }

  $scope.isOrderSelected = () => $scope.getActiveOrders().nonEmpty && $scope.selectedOrder.nonEmpty

  $scope.popupNewOrderDialog = (anAccountType: js.UndefOr[String]) => anAccountType map { accountType =>
    newOrderDialog.popup(NewOrderParams(
      symbol = $cookies.getOrElse(DiscoverController.LastSymbolCookie, "AAPL"),
      accountType = accountType
    ))
  }

  $scope.selectOrder = (order: js.UndefOr[Order]) => $scope.selectedOrder = order

  $scope.toggleSelectedOrder = () => $scope.selectedOrder = js.undefined

  /////////////////////////////////////////////////////////////////////
  //          Performance Functions
  /////////////////////////////////////////////////////////////////////

  $scope.getPerformance = () => mySession.getPerformance

  $scope.isPerformanceSelected = () => $scope.getPerformance().nonEmpty && $scope.selectedPerformance.nonEmpty

  $scope.selectPerformance = (performance: js.UndefOr[Performance]) => $scope.selectedPerformance = performance

  $scope.toggleSelectedPerformance = () => $scope.selectedPerformance = js.undefined

  $scope.cost = (aTx: js.UndefOr[Performance]) => aTx.flatMap(_.totalCost)

  $scope.soldValue = (aTx: js.UndefOr[Performance]) => aTx.flatMap(_.totalSold)

  $scope.proceeds = (aTx: js.UndefOr[Performance]) => aTx.flatMap(_.proceeds)

  $scope.gainLoss = (aTx: js.UndefOr[Performance]) => aTx.flatMap(_.gainLoss)

  /////////////////////////////////////////////////////////////////////
  //          Position Functions
  /////////////////////////////////////////////////////////////////////

  $scope.getPositions = () => {
    mySession.participant foreach enrichPositions
    mySession.getPositions filter (_.accountType == $scope.getAccountType())
  }

  $scope.isPositionSelected = () => $scope.getPositions().nonEmpty && $scope.selectedPosition.nonEmpty

  $scope.selectPosition = (position: js.UndefOr[Position]) => $scope.selectedPosition = position

  $scope.sellPosition = (aSymbol: js.UndefOr[String], aQuantity: js.UndefOr[Double]) => {
    for {
      symbol <- aSymbol
      quantity <- aQuantity
    } yield newOrderDialog.popup(NewOrderParams(symbol = symbol, quantity = quantity))
  }

  $scope.toggleSelectedPosition = () => $scope.selectedPosition = js.undefined

  $scope.tradingStart = () => new js.Date()

  /////////////////////////////////////////////////////////////////////
  //          Private Functions
  /////////////////////////////////////////////////////////////////////

  private def enrichOrders(participant: Participant) {
    if (mySession.participant.nonEmpty) {
      if (!isDefined(participant.dynamic.enrichedOrders)) {
        participant.dynamic.enrichedOrders = true

        for {
          contestId <- mySession.getContestID
          playerId <- participant._id
        } {
          contestService.getEnrichedOrders(contestId, playerId) onComplete {
            case Success(enrichedOrders) => mySession.participant.foreach(_.orders = enrichedOrders)
            case Failure(e) =>
              toaster.error("Error!", "Error loading enriched orders")
          }
        }
      }
    }
  }

  private def enrichPositions(participant: Participant) {
    if (mySession.participant.nonEmpty) {
      if (!isDefined(participant.dynamic.enrichedPositions)) {
        participant._id foreach { playerId =>
          participant.dynamic.enrichedPositions = true
          for {
            contestId <- mySession.getContestID
          } {
            contestService.getEnrichedPositions(contestId, playerId) onComplete {
              case Success(enrichedPositions) => mySession.participant.foreach(_.positions = enrichedPositions)
              case Failure(e) => toaster.error("Error loading enriched positions")
            }
          }
        }
      }
    }
  }

  private def resetOrders() {
    $scope.selectedOrder = js.undefined
    $scope.selectedClosedOrder = js.undefined
  }

  private def resetPositions() {
    $scope.selectedPosition = js.undefined
    $scope.selectedPerformance = js.undefined
  }

  //////////////////////////////////////////////////////////////////////
  //              Watch Event Listeners
  //////////////////////////////////////////////////////////////////////

  $scope.$on(ContestSelected, { (event: js.Dynamic, contest: js.Dynamic) =>
    console.log(s"[Portfolio] Contest '${contest.name}' selected")
    resetOrders()
    resetPositions()
  })

  $scope.$on(ContestUpdated, { (event: js.Dynamic, contest: js.Dynamic) =>
    console.log(s"[Portfolio] Contest '${contest.name}' updated")
    resetOrders()
    resetPositions()
  })

  $scope.$on(OrderUpdated, { (event: js.Dynamic, contest: js.Dynamic) =>
    console.log(s"[Portfolio] Orders for Contest '${contest.name}' updated")
    resetOrders()
  })

  $scope.$on(ParticipantUpdated, { (event: js.Dynamic, contest: js.Dynamic) =>
    console.log(s"[Portfolio] Orders for Contest '${contest.name}' updated")
    resetPositions()
  })

}

/**
  * Portfolio Scope
  * @author lawrence.daniels@gmail.com
  */
@js.native
trait PortfolioScope extends Scope {
  // variables
  var portfolioTabs: js.Array[PortfolioTab] = js.native
  var selectedClosedOrder: js.UndefOr[ClosedOrder] = js.native
  var selectedOrder: js.UndefOr[Order] = js.native
  var selectedPerformance: js.UndefOr[Performance] = js.native
  var selectedPosition: js.UndefOr[Position] = js.native

  // closed order functions
  var getClosedOrders: js.Function0[js.Array[ClosedOrder]] = js.native
  var isClosedOrderSelected: js.Function0[Boolean] = js.native
  var selectClosedOrder: js.Function1[js.UndefOr[ClosedOrder], Unit] = js.native
  var toggleSelectedClosedOrder: js.Function0[Unit] = js.native

  // order functions
  var computeOrderCost: js.Function1[js.UndefOr[Order], js.UndefOr[Double]] = js.native
  var cancelOrder: js.Function3[js.UndefOr[String], js.UndefOr[String], js.UndefOr[String], Unit] = js.native
  var getActiveOrders: js.Function0[js.Array[Order]] = js.native
  var getAccountType: js.Function0[String] = js.native
  var isMarketOrder: js.Function1[js.UndefOr[Order], Boolean] = js.native
  var isOrderSelected: js.Function0[Boolean] = js.native
  var selectOrder: js.Function1[js.UndefOr[Order], Unit] = js.native
  var popupNewOrderDialog: js.Function1[js.UndefOr[String], js.UndefOr[js.Promise[NewOrderDialogResult]]] = js.native
  var toggleSelectedOrder: js.Function0[Unit] = js.native

  // performance functions
  var cost: js.Function1[js.UndefOr[Performance], js.UndefOr[Double]] = js.native
  var gainLoss: js.Function1[js.UndefOr[Performance], js.UndefOr[Double]] = js.native
  var proceeds: js.Function1[js.UndefOr[Performance], js.UndefOr[Double]] = js.native
  var soldValue: js.Function1[js.UndefOr[Performance], js.UndefOr[Double]] = js.native
  var getPerformance: js.Function0[js.Array[Performance]] = js.native
  var isPerformanceSelected: js.Function0[Boolean] = js.native
  var selectPerformance: js.Function1[js.UndefOr[Performance], Unit] = js.native
  var toggleSelectedPerformance: js.Function0[Unit] = js.native

  // position functions
  var getPositions: js.Function0[js.Array[Position]] = js.native
  var isPositionSelected: js.Function0[Boolean] = js.native
  var selectPosition: js.Function1[js.UndefOr[Position], Unit] = js.native
  var sellPosition: js.Function2[js.UndefOr[String], js.UndefOr[Double], js.UndefOr[js.Promise[NewOrderDialogResult]]] = js.native
  var toggleSelectedPosition: js.Function0[Unit] = js.native
  var tradingStart: js.Function0[js.Date] = js.native

}

/**
  * Portfolio Tab
  * @author lawrence.daniels@gmail.com
  */
@ScalaJSDefined
class PortfolioTab(val name: String, val icon: String, val path: String, val active: Boolean = false) extends js.Object
